src/client/pfs_file.go

```diff
@@ -123,38 +125,35 @@ type modifyFileCore struct {
 }
 
 func (mfc *modifyFileCore) PutFile(path string, r io.Reader, opts ...PutFileOption) error {
+	config := &putFileConfig{}
+	for _, opt := range opts {
+		opt(config)
+	}
 	return mfc.maybeError(func() error {
-		pf := &pfs.PutFile{
-			Source: &pfs.PutFile_RawFileSource{
-				RawFileSource: &pfs.RawFileSource{
-					Path: path,
-				},
-			},
-		}
-		for _, opt := range opts {
-			opt(pf)
+		if !config.append {
+			if err := mfc.sendDeleteFile(&pfs.DeleteFile{
+				Path: path,
+				Tag:  config.tag,
+			}); err != nil {
+				return err
+			}
 		}
-		if err := mfc.sendPutFile(pf); err != nil {
+		if err := mfc.sendPutFile(&pfs.AddFile{
```

Comment:
This is still operating on a stream. I think you should have an add file message that is repeatedly updated and sent with the data for each chunk. We are not operating on the messages independently with this approach.

Reply:
I think that it's probably a good idea to assume the last path so we don't have to keep resending it.  This will come in hand especially for large files.  The cost in server implementation complexity is 2 if statements.

Reply:
Resending the path is negligible when we can fit ~20MB of data in each message, and it is simple on the client side and simpler on the server side. I'm surprised you're not in favor of this because it will make it much easier to ensure that each message in the stream is setup properly since we can validate them in isolation.

Reply:
Okay, will remove.